core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala

```diff
@@ -303,6 +303,32 @@ class StandaloneDynamicAllocationSuite
     assert(master.apps.head.getExecutorLimit === 1)
   }
 
+   test("the pending replacement executors should not be lost (SPARK-10515)") {
+    sc = new SparkContext(appConf)
+    val appId = sc.applicationId
+    assert(master.apps.size === 1)
+    assert(master.apps.head.id === appId)
+    assert(master.apps.head.executors.size === 2)
+    assert(master.apps.head.getExecutorLimit === Int.MaxValue)
+    // sync executors between the Master and the driver, needed because
+    // the driver refuses to kill executors it does not know about
+    syncExecutors(sc)
+    val executors = getExecutorIds(sc)
+    assert(executors.size === 2)
+
+    // kill executor,and replace it
+    assert(sc.killAndReplaceExecutor(executors.head))
```

Comment:
Hmm... this doesn't look right. It should probably be using a new version of `killNExecutors` that calls `killAndReplaceExecutor` and syncs things.

@andrewor14 might be a better person to comment on this, since he wrote the original tests. I'm not sure about how much we can trust the counts to update atomically when `killNExecutors` and friends are called, but the other tests seem to be passing reliably...